Places/View Controllers/ShortestRouteMapViewController.m

```diff
@@ -0,0 +1,190 @@
+    NSLog(@"lat: %f lng: %f", lat, lng);
+    GMSCameraPosition *camera = [GMSCameraPosition cameraWithLatitude:lat
+                                                            longitude:lng
+                                                                 zoom:12];
+    self.mapView.camera = camera;
+}
+
+- (void)getStartingWaypointsEndingParameters {
+    NSArray *placesToGo = self.itinerary.placesToGo;
+    NSMutableArray *placesToGoIDs = [[NSMutableArray alloc] init];
+    for (Place *place in placesToGo) {
+        place.fetchIfNeeded;
+        [placesToGoIDs addObject:place.placeID];
+    }
+    NSLog(@"This is the array of places to go IDs %@", placesToGoIDs);
+    
+    for (NSInteger i=0; i < [placesToGoIDs count]; i++) {
+        NSString *curPlaceID = placesToGoIDs[i];
+        if (i == 0) {
+            self.originParameter = [NSString stringWithFormat:@"place_id:%@", curPlaceID];
+        } else if (i == (placesToGo.count - 1)) {
+            self.destinationParameter = [NSString stringWithFormat:@"place_id:%@", curPlaceID];
+        } else { // one of the waypoints
+            if (self.waypointsParameter) {
+                NSString *toAppend = [NSString stringWithFormat:@"|place_id:%@", curPlaceID];
+                [self.waypointsParameter appendString:toAppend];
+            } else { // first waypoint, don't need pipe
+                self.waypointsParameter = [NSMutableString stringWithFormat:@"place_id:%@", curPlaceID];
+            }
+        }
+    }
+    NSLog(@"waypointsParameter: %@", self.waypointsParameter);
+    NSLog(@"originParameter: %@", self.originParameter);
+    NSLog(@"destinationParameter: %@", self.destinationParameter);
+}
+
+- (void)requestRouteToDraw {
+    NSString *urlString = [NSString stringWithFormat: @"%@?origin=%@&destination=%@&waypoints=%@&sensor=true&mode=%@&key=%@",
+                           @"https://maps.googleapis.com/maps/api/directions/json",
+                           self.originParameter,
+                           self.destinationParameter,
+                           self.waypointsParameter,
+                           self.selectedTravelMode,
+                           @"AIzaSyA2kTwxS9iiwWd3ydaxxwdewfAjZdKJeDE"];
+    NSLog(@"This is the urlString %@", urlString);
+    urlString = [urlString stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
+    NSURL *directionsURL = [NSURL URLWithString:urlString];
+    NSLog(@"This is the request url %@", directionsURL);
+    ASIHTTPRequest *request = [ASIHTTPRequest requestWithURL:directionsURL];
+    [request startSynchronous];
+    NSError *error = [request error];
+    if (!error) {
+        NSString *response = [request responseString];
+        NSLog(@"%@",response);
+        NSDictionary *json =[NSJSONSerialization JSONObjectWithData:[request responseData] options:NSJSONReadingMutableContainers error:&error];
+        GMSPath *path =[GMSPath pathFromEncodedPath:json[@"routes"][0][@"overview_polyline"][@"points"]];
```

Comment:
I think I've mentioned this to you in one of the CodePath assignments, but you need to be very careful when dealing with dictionaries parsed from a JSON that you have no control over the contents. You shouldn't make any assumptions that the data will be formatted in a specific way. JSON doesn't have a schema, so you need to do post-parsing validation instead.

For instance, in this particular case, you're making the following assumptions about the JSON sent by the server:

1. it will contain an attribute `routes`
2. `routes` will be an array with least one item
3. the first item in the `routes` array is going to be a dictionary
4. that dictionary will contain an attribute called `overview_polyline`
5. the value of `overview_polyline` will be a dictionary
6. that dictionary will contain an attribute called `points`
7. the value of `points` will be an encoded `GMSPath`

Your app might crash if any of those assumptions isn't true. Maybe the server returns an error? Maybe the API gets updated? (unlikely, since it should be versioned, but still worth considering)

You can workaround these problems by implementing a more robust parsing logic that doesn't make any assumptions and, instead, validates the data at runtime.

Tip: I've found that adding the following extension to `NSObject` makes type checking/casting a bit simpler:
```
@interface NSObject (SafeCast)
+ (nullable instancetype)castFrom:(id)object;
@end
@implementation NSObject (SafeCast)
+ (nullable instancetype)castFrom:(id)object {
   return [object isKindOfClass:self] ? object : nil;
}
@end
```